### PR TITLE
 sys/shell: cancel current line on CTRL-C.

### PR DESCRIPTION
**Pull request by jcarrano**
### Contribution description

Just like in a regular UNIX terminal, control-C now cancels the current line.  This is a suggestion of @benemorius.

This is useful if one is using a dumb terminal to communicate with a node, as it saves having to repeatedly type backspace to discard the current line. It also helps when connecting to an already running node,
as one does not know what is on the line buffer, the safest thing to do is to begin by sending a ctrl-C.

#### Other changes

To be able to test this I had to do a couple of modifications to `tests/shell`.

- Fix the use of python dictionaries: Python dictionaries are not guaranteed to be ordered until version
3.7. In 3.6 they are ordered too, but that is an implementation detail. riotdocker seems to be using 3.5. The commands in this test were stored in a dict.

- Use miniterm instead of pyterm: pyterm messes up control characters, so the test was changed to use
miniterm.py instead.

### Testing procedure

I have included a test. It should work both on a board or in native:

```
$ cd tests/shell
$ BOARD=samr21-xpro make all
$ BOARD=samr21-xpro make flash
$ BOARD=samr21-xpro make test
```

If testing by hand:

- On a board, be sure to use a terminal that forwards ctrl-C, otherwise you will not be testing the thing. This means _no_ pyterm.
- On native, use Ctrl-V+ctrl-C to insert a literal ETX, otherwise you will be sending a SIGINT.

### Issues/PRs references

Depends on ~#11003~ for testing.
Part of #10994 .
Split from #10788 .

### CODE REVIEW

**Comment by miri64** [comment]
Doesn't work on native ;-)

**Comment by jcarrano** [comment]
> Doesn't work on native ;-)

It works if you do Ctrl-V Ctrl-C, which is what the test script does. Otherwise the native port catches it as a SIGINT and the process exits. That can be changed if it makes sense.

**Comment by miri64** [comment]
Nope, when I try it in `examples/gnrc_networking` with `ping6 -c 10000 fe80::...` (where the address is the address of another node or the TAP bridge), I'm not able to cancel it.

File: tests/shell/Makefile [hunk 1]
```diff
@@ -12,4 +13,8 @@ BOARD_BLACKLIST += chronos
 
 TEST_ON_CI_WHITELIST += all
 
+ifneq ($(BOARD), native)
+	RIOT_TERMINAL=miniterm.py
```
Comment (kaspar030) [inline tests/shell/Makefile]: make this conditional. CI uses picocom.

**Comment by benemorius** [comment]
> Nope, when I try it in `examples/gnrc_networking` with `ping6 -c 10000 fe80::...` (where the address is the address of another node or the TAP bridge), I'm not able to cancel it.

I didn't test this yet but it sounds like you're thinking this should terminate a running process with ^C (that's what #10715 is about) which would be lovely too, but this PR is instead about clearing the command you're typing to get a fresh/empty prompt to replace one that has some partial or wrong command typed in. It's a little thing but it's a very nice thing because the shells we're all accustomed to using on *nix behave this way so it improves our shell UX considerably.

In practice it looks like this:
```
> pin6 26^C
> ping6 2600::
...
```

**Comment by jcarrano** [comment]
I'm realising now that this is __not__ the best way to implement this feature (it is now wrong either.)

The right thing to do here would be to implement a sort of ["line discipline"](http://www.linusakesson.net/programming/tty/index.php) in `stdio_uart`. It would be the correct place to implement:

- echoing
- line buffering
- dropping the line on ctrl-c (or sending a message/flag to some "foreground thread". I'm not saying it should do that, but if someone wants the feature like @benemorius then this is the place to put it.)
- [Sending a partially completed line (even an empty one) on ctrl-D](https://utcc.utoronto.ca/~cks/space/blog/unix/TypingEOFEffects).

Of course, `stdio_uart` would get bigger and more complex but it would probably get offset by the decrease in complexity of the shell code that handles this at the moment. In addition to that, the functionality would be available to any shell command (for free) and if it is considered too heavy, an alternative implementation of `stdio_uart` without the magic can be provided.

By the way, the line discipline is the place to put stuff like ethos/slip (this is what's done in unix) though ideally in a separate module and not patched with #ifdefs.

**Comment by miri64** [comment]
> but it sounds like you're thinking this should terminate a running process with ^C (that's what #10715 is about) which would be lovely too, but this PR is instead about clearing the command you're typing to get a fresh/empty prompt to replace one that has some partial or wrong command typed in.

Yepp, I misunderstood the purpose of this PR. Sorry for the noise!

**Comment by jcarrano** [comment]
Now that socat is available as a RIOT_TERMINAL it rebased and used it in the test.

**Comment by cladmi** [comment]
This breaks on boards that cannot run when configured with `RIOT_TERMINAL != pyterm`

```
git grep TERMFLAGS boards/common/msba2/ boards/lobaro-lorabox/
boards/common/msba2/Makefile.include:TERMFLAGS += -tg -p "$(PORT)"
boards/lobaro-lorabox/Makefile.include:TERMFLAGS +=  --set-rts 0
```

**Comment by cladmi** [comment]
If it expects a clean terminal output, it will, I think, also fail on all boards using `stdio_rtt`.

**Pull request merged**